docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,140 @@
+In case of services in a Kubernetes cluster, we don't have to do anything for local services because we can just rely on Kube DNS that resolves to a Kube VIP we placed for MeshService.
+
+In case of services in a Universal cluster, currently we provide two ways:
+* `<kuma.io/service>.mesh` address that resolves to Kuma VIP
+* VirtualOutbound resource to provide custom addresses
+
+Hostnames should be managed only by MeshOperator. Otherwise, we get into security issues such as one service owner can override a hostname of other service.
+Hostnames are created by convention in the majority of cases - Kubernetes hostnames are also such example.
+It is still possible for service owner to create unique hostname for their app by reaching out to mesh operator.
+
+To manage hostname, we introduce new object called `HostnameGenerator`
+
+```yaml
+type: HostnameGenerator
+name: k8s-zone-hostnames
+spec:
+  selector:
+    meshService:
+      Labels:
+        kuma.io/zone: east 
+  template: {{ label "k8s.kuma.io/service-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.mesh.{{ label "kuma.io/zone" }}
+```
+
+`HostnameGenerator` is a namespaced-scoped object on Kubernetes, but can only be applied in system namespace.
+It is namespaced-scoped object, because we already learned that it's better to start with restricted namespace object than to start with cluster-scope and then make it namespaced-scope.
+We might need this for deploying Global and Zone CP in one Kube cluster and maybe potentially enable applying generator on namespaces to affect only data plane proxes in specific namespace.
+It can be applied on global CP, so it's synced to all zones, or it can be applied on specific zones.
+
+For example, given this mesh service
+
+```yaml
+type: MeshService
+name: redis.demo-app
+labels:
+  k8s.kuma.io/service-name: redis
+  k8s.kuma.io/namespace: demo-app
+  kuma.io/zone: east
+spec: ...
+```
+We would generate such hostname `redis.demo-app.svc.mesh.east`
+
+Possible template functions/keys are:
+* `{{ name }}` - name of the MeshService
+* `{{ label "x" }}` - value of label `x`.
+
+If the template cannot be resolved (label is missing), the hostname won't be generated with a `NotAvailable` status and an appropriate `reason`.
+
+Generated hostnames will be placed in the status of MeshService
+```yaml
+status:
+  addresses:
+    - hostname: redis.demo-app.svc.mesh.east
+      status: Available # | NotAvailable if there was a collision
+      origin:
+        kind: HostnameGenerator
+        name: "k8s-zone-hostnames" # name of the HostnameGenerator
+      reason: "not available because of the clash with ..." # reason if there was a problem
+```
+
+Hostname generation process (only computed by leader):
+* List MeshService and HostnameGenerators
+* Sort HostnameGenerators by precedence
+* Take each HostnameGenerator and add hostnames for a Service (on the side, do not modify MeshService object). Track generated hostnames to detect collisions 
```

Comment:
What about a case where we have 2 services in 2 different mesh but in the same namespace? In this case, we might not have a clash of IP addresses but hostnames.

Reply:
> What about a case where we have 2 services in 2 different mesh but in the same namespace?

How to define this using Kubernetes objects? Can you give an example?

Reply:
True, it's not possible but is it possible in universal?

Reply:
Ok, true. This process should be sharded by mesh. I mean this step "Track generated hostnames to detect collisions" should be per mesh. Let me update the MADR